### PR TITLE
🎨 Cleaner parser tests

### PR DESCRIPTION
**Pull request by anzdaddy**
Changes proposed in this pull request:
- Use computed temp filenames instead of g{olden,enerated}.txt.
- Enable golden-output regression tests.

Builds on yet-to-be-merged #168. Only commits after 979fc8c are specific to this PR.

@anz-bank/sysl-developers

### CODE REVIEW

File: sysl2/sysl/Parser_test.go [hunk 1]
```diff
@@ -70,36 +78,71 @@ func parseComparable(filename, root string) (*sysl.Module, error) {
...
+	}
+
+	generated, err := ioutil.TempFile("", "sysl-test-*.textpb")
 	if err != nil {
 		return false, err
 	}
-	if proto.Equal(pyModule, goModule) {
-		return true, nil
-	}
+	defer generated.Close()
```
Comment (anzboi) [inline sysl2/sysl/Parser_test.go]: `generated.Close()` is called further down on line 105. The deferred call is generating an error but is ignored due to defer.

File: sysl2/sysl/Parser_test.go [hunk 2]
```diff
@@ -111,171 +154,181 @@ func parseAndPrint(t *testing.T, filename, root string) error {
 	return nil
 }
 
-func testParse(t *testing.T, filename, root string) {
-	equal, err := parseAndCompare(filename, root)
+func testParseAgainstPython(t *testing.T, filename, root string) {
```
Comment (anzboi) [inline sysl2/sysl/Parser_test.go]: root is only non-empty in two of the tests below (lines 263 and 329). In all other cases, the filename is given with the path. Is root something we want to keep for future or do we get want to get rid of it?

File: sysl2/sysl/Parser_test.go [hunk 3]
```diff
@@ -111,171 +154,181 @@ func parseAndPrint(t *testing.T, filename, root string) error {
...
+	equal, err := parseAndCompareWithPython(filename, root)
+	assert.NoError(t, err)
+	assert.True(t, equal, "Mismatch")
+}
+
+func testParseAgainstGolden(t *testing.T, filename, root string) {
```
Comment (anzboi) [inline sysl2/sysl/Parser_test.go]: Similar here, except root is never non-empty.

File: sysl2/sysl/Parser_test.go (hunk 2 above)
Comment (sahejsingh) [reply sysl2/sysl/Parser_test.go]: root is required for import statements to work.

File: sysl2/sysl/Parser_test.go [hunk 4]
```diff
@@ -13,39 +14,46 @@ import (
...
+
+	module := &sysl.Module{}
+	if err := proto.UnmarshalText(buf.String(), module); err != nil {
+		return nil, err
+	}
+	return module, nil
+}
+
+func pySysl() string {
+	if pySysl, found := os.LookupEnv("SYSL_PYTHON_BIN"); found {
```
Comment (juliaogris) [inline sysl2/sysl/Parser_test.go]: Nit: I believe go convention is to call the var above `ok` not `found` (see [docs](https://golang.org/pkg/os/#LookupEnv))

File: sysl2/sysl/Parser_test.go [hunk 5]
```diff
@@ -13,39 +14,46 @@ import (
...
+	return "sysl"
+}
+
+func pyParse(filename, root, output string) (*sysl.Module, error) {
+	args := []string{"textpb", "-o", output, filename}
 	if len(root) > 0 {
 		rootArg := []string{"--root", root}
-		// TODO: This looks dubious
-		args[2] = root + "/" + args[2]
 		output = args[2]
```
Comment (juliaogris) [inline sysl2/sysl/Parser_test.go]: Remove?

File: sysl2/sysl/Parser_test.go [hunk 6]
```diff
@@ -70,36 +78,71 @@ func parseComparable(filename, root string) (*sysl.Module, error) {
 	return module, nil
 }
 
-func parseAndCompare(filename, root string) (bool, error) {
-	goModule, err := parseComparable(filename, root)
+func parseAndCompare(filename, root, golden string, goldenModule *sysl.Module) (bool, error) {
+	module, err := parseComparable(filename, root)
```
Comment (juliaogris) [inline sysl2/sysl/Parser_test.go]: Not part of this PR, but seeing as it's a cleanup: remove line 69 above (`// app.SourceContext.Start.Col = 0`)

File: sysl2/sysl/Parser_test.go [hunk 7]
```diff
@@ -70,36 +78,71 @@ func parseComparable(filename, root string) (*sysl.Module, error) {
...
 
-	golden := "golden.txt"
-	generated := "generated.txt"
-	if err = TextPB(pyModule, golden); err != nil {
+	if err = TextPB(goldenModule, golden); err != nil {
 		return false, err
 	}
-	if err = TextPB(goModule, generated); err != nil {
+
+	if err = FTextPB(generated, module); err != nil {
```
Comment (juliaogris) [inline sysl2/sysl/Parser_test.go]: `generated` (tempfile) first used here, define right above (I think)

File: sysl2/sysl/Parser_test.go [hunk 8]
```diff
@@ -70,36 +78,71 @@ func parseComparable(filename, root string) (*sysl.Module, error) {
...
+
+	generated, err := ioutil.TempFile("", "sysl-test-*.textpb")
 	if err != nil {
 		return false, err
 	}
-	if proto.Equal(pyModule, goModule) {
-		return true, nil
-	}
+	defer generated.Close()
+	defer os.Remove(generated.Name())
```
Comment (juliaogris) [inline sysl2/sysl/Parser_test.go]: Are you sure you want to remove the tempfile even if the tests fails? Might be worth keeping it around for investigation.

File: sysl2/sysl/Parser_test.go (hunk 1 above)
Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: Both calls are needed and it's easier to call `Close` redundantly than to check whether it has already been called. I could add a check or simply an explanatory comment.

Comment (sahejsingh) [reply sysl2/sysl/Parser_test.go]: uh. more complicated code with defer. reminds me of c++ destructors/RAII. Goes to show error handling is not that simple. You end up having multiple code paths to think about.

@marcelocantos Is the intent of `defer close` to close the file in case of returning from  99 and 103 and the close at 103 is to make sure its closed before diff?

Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: Done (added a check).

File: sysl2/sysl/Parser_test.go (hunk 8 above)
Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: Good idea! Done.

File: sysl2/sysl/Parser_test.go (hunk 7 above)
Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: Done.

File: sysl2/sysl/Parser_test.go (hunk 3 above)
Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: As above.

File: sysl2/sysl/Parser_test.go (hunk 7 above)
Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: I've also done it in `parseAndCompareWithPython`.

File: sysl2/sysl/Parser_test.go (hunk 4 above)
Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: Done. I also reworked it into a computed-once variable instead of a function call.

File: sysl2/sysl/Parser_test.go (hunk 5 above)
Comment (juliaogris) [reply sysl2/sysl/Parser_test.go]: Sorry, I must be having a completely dumb moment, but isn't `output` already equal to `args[2]`

Comment (anzdaddy) [reply sysl2/sysl/Parser_test.go]: Ah yes, I see now. Off-by-one error in my brain.
